src/game/room.c

```diff
@@ -590,19 +590,52 @@ int16_t Room_GetWaterHeight(int32_t x, int32_t y, int32_t z, int16_t room_num)
 
 void Room_AlterFloorHeight(ITEM_INFO *item, int32_t height)
 {
-    int16_t room_num = item->room_number;
-    FLOOR_INFO *floor =
-        Room_GetFloor(item->pos.x, item->pos.y, item->pos.z, &room_num);
-    FLOOR_INFO *ceiling = Room_GetFloor(
-        item->pos.x, item->pos.y + height - WALL_L, item->pos.z, &room_num);
-
-    if (floor->floor == NO_HEIGHT / 256) {
-        floor->floor = ceiling->ceiling + height / 256;
-    } else {
-        floor->floor += height / 256;
-        if (floor->floor == ceiling->ceiling && floor->sky_room == NO_ROOM) {
+    if (!height) {
+        return;
+    }
+
+    int16_t data;
+    FLOOR_INFO *floor;
+    ROOM_INFO *r = &g_RoomInfo[item->room_number];
+
+    do {
+        int32_t z_floor = (item->pos.z - r->z) >> WALL_SHIFT;
+        int32_t x_floor = (item->pos.x - r->x) >> WALL_SHIFT;
+
+        if (z_floor <= 0) {
+            z_floor = 0;
+            if (x_floor < 1) {
+                x_floor = 1;
+            } else if (x_floor > r->y_size - 2) {
+                x_floor = r->y_size - 2;
+            }
+        } else if (z_floor >= r->x_size - 1) {
+            z_floor = r->x_size - 1;
+            if (x_floor < 1) {
+                x_floor = 1;
+            } else if (x_floor > r->y_size - 2) {
+                x_floor = r->y_size - 2;
+            }
+        } else if (x_floor < 0) {
+            x_floor = 0;
+        } else if (x_floor >= r->y_size) {
+            x_floor = r->y_size - 1;
+        }
+
+        floor = &r->floor[z_floor + x_floor * r->x_size];
+        data = Room_GetDoor(floor);
+        if (data != NO_ROOM) {
+            r = &g_RoomInfo[data];
+        }
+    } while (data != NO_ROOM);
+
+    if (floor->floor != NO_HEIGHT / 256) {
```

Comment:
Should we update to use `STEP_L` in place of `256`?

Reply:
I will make a second, separate commit for this because `NO_HEIGHT / 256` is used in other spots like `Door_Shut` and `MovableBlock_TestDestination`.

Reply:
It might be good to standardize throughout the codebase as sometimes you see `h / 256` and other times `h >> 8`. So maybe that's one to tackle under a separate issue.

Reply:
Alright I opened #1053 for this.